File: cmd/forged/middleware/rate_limit.go
package middleware

import (
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"strings"
	"sync"
	"time"
)

// RateLimiter provides configurable rate limiting with per-endpoint support
type RateLimiter struct {
	clients     map[string]*clientStats
	mu          sync.RWMutex
	defaultLimit int
	windowSize  time.Duration
	// Per-endpoint limits: path -> requests per window
	endpointLimits map[string]int
	// Skip list: paths that bypass rate limiting
	skipPaths map[string]bool
	// Enable API token extraction from headers
	tokenHeader string
}

// clientStats tracks request count and window for a client
type clientStats struct {
	hits       int
	lastWindow time.Time
}

// RateLimitConfig allows customization of the rate limiter
type RateLimitConfig struct {
	DefaultLimit   int
	WindowSize     time.Duration
	EndpointLimits map[string]int // path -> limit (0 = use default)
	SkipPaths      []string       // paths to skip rate limiting
	TokenHeader    string         // header name for API token (e.g., "X-API-Token")
}

// DefaultRateLimiter is the global rate limiter instance
var DefaultRateLimiter = NewRateLimiter(&RateLimitConfig{
	DefaultLimit: 100,
	WindowSize:   time.Minute,
	EndpointLimits: map[string]int{
		"/api/health":      0, // Use default (no special limit)
		"/health":          0,
		"/api/status":      200, // Higher limit for status
		"/metrics":         60,  // Lower limit for metrics
		"/api/tui/logs":    300, // Higher for dashboard
		"/api/tui/dashboard": 300,
	},
	SkipPaths:   []string{"/health", "/api/health"},
	TokenHeader: "X-API-Token",
})

// NewRateLimiter creates a new rate limiter with the given configuration
func NewRateLimiter(cfg *RateLimitConfig) *RateLimiter {
	if cfg.DefaultLimit <= 0 {
		cfg.DefaultLimit = 100
	}
	if cfg.WindowSize <= 0 {
		cfg.WindowSize = time.Minute
	}

	rl := &RateLimiter{
		clients:        make(map[string]*clientStats),
		defaultLimit:   cfg.DefaultLimit,
		windowSize:     cfg.WindowSize,
		endpointLimits: cfg.EndpointLimits,
		skipPaths:      make(map[string]bool),
		tokenHeader:    cfg.TokenHeader,
	}

	for _, path := range cfg.SkipPaths {
		rl.skipPaths[path] = true
	}

	return rl
}

// getClientIdentifier extracts a unique identifier for the client
// Priority: API token > IP address
func (rl *RateLimiter) getClientIdentifier(r *http.Request) string {
	// Check for API token first
	if rl.tokenHeader != "" {
		if token := r.Header.Get(rl.tokenHeader); token != "" {
			return "token:" + token
		}
	}

	// Fall back to IP address
	ip, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		ip = r.RemoteAddr
	}

	// Handle X-Forwarded-For if behind proxy
	if forwarded := r.Header.Get("X-Forwarded-For"); forwarded != "" {
		// Use the first IP in the chain (client IP)
		ips := strings.Split(forwarded, ",")
		if len(ips) > 0 {
			ip = strings.TrimSpace(ips[0])
		}
	}

	return "ip:" + ip
}

// shouldSkip checks if the request path should skip rate limiting
func (rl *RateLimiter) shouldSkip(path string) bool {
	rl.mu.RLock()
	defer rl.mu.RUnlock()

	// Exact match
	if rl.skipPaths[path] {
		return true
	}

	// Prefix match for paths like /api/health/detailed
	for skipPath := range rl.skipPaths {
		if strings.HasPrefix(path, skipPath+"/") {
			return true
		}
	}

	return false
}

// getLimitForPath returns the rate limit for a specific path
func (rl *RateLimiter) getLimitForPath(path string) int {
	rl.mu.RLock()
	defer rl.mu.RUnlock()

	// Check for exact match first
	if limit, ok := rl.endpointLimits[path]; ok {
		if limit > 0 {
			return limit
		}
		// limit == 0 means use default
	}

	// Check for prefix matches (e.g., /api/tasks/123/events matches /api/tasks)
	for endpoint, limit := range rl.endpointLimits {
		if strings.HasPrefix(path, endpoint+"/") && limit > 0 {
			return limit
		}
	}

	return rl.defaultLimit
}

// isAllowed checks if a request from the given client is allowed
// Returns (allowed, currentHits, limit, resetTime)
func (rl *RateLimiter) isAllowed(clientID, path string) (bool, int, int, time.Time) {
	// Get limit first (needs read lock)
	limit := rl.getLimitForPath(path)

	rl.mu.Lock()
	defer rl.mu.Unlock()

	now := time.Now()
	window := now.Truncate(rl.windowSize)
	resetTime := window.Add(rl.windowSize)

	// Create composite key: clientID + path for per-endpoint limiting
	key := clientID + ":" + path

	stats, exists := rl.clients[key]
	if !exists || stats.lastWindow.Before(window) {
		stats = &clientStats{
			hits:       0,
			lastWindow: window,
		}
		rl.clients[key] = stats
	}

	stats.hits++
	return stats.hits <= limit, stats.hits, limit, resetTime
}

// Middleware returns the rate limiting middleware function
func (rl *RateLimiter) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Check if path should skip rate limiting
		if rl.shouldSkip(r.URL.Path) {
			next.ServeHTTP(w, r)
			return
		}

		clientID := rl.getClientIdentifier(r)
		allowed, hits, limit, resetTime := rl.isAllowed(clientID, r.URL.Path)

		remaining := limit - hits
		if remaining < 0 {
			remaining = 0
		}

		// Set rate limit headers
		w.Header().Set("X-RateLimit-Limit", fmt.Sprintf("%d", limit))
		w.Header().Set("X-RateLimit-Remaining", fmt.Sprintf("%d", remaining))
		w.Header().Set("X-RateLimit-Reset", fmt.Sprintf("%d", resetTime.Unix()))

		if !allowed {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusTooManyRequests)
			json.NewEncoder(w).Encode(map[string]interface{}{
				"error":       "rate limit exceeded",
				"code":        "RATE_LIMIT_EXCEEDED",
				"limit":       limit,
				"window":      rl.windowSize.String(),
				"retry_after": int(time.Until(resetTime).Seconds()),
			})
			return
		}

		next.ServeHTTP(w, r)
	})
}

// RateLimitMiddleware is the legacy middleware function using the default rate limiter
// It provides backward compatibility while using the new configurable backend
func RateLimitMiddleware(next http.Handler) http.Handler {
	return DefaultRateLimiter.Middleware(next)
}

// WithRateLimitConfig creates a new middleware with custom configuration
func WithRateLimitConfig(cfg *RateLimitConfig) func(http.Handler) http.Handler {
	rl := NewRateLimiter(cfg)
	return rl.Middleware
}

// SkipPaths returns the paths that are configured to skip rate limiting
func (rl *RateLimiter) SkipPaths() []string {
	rl.mu.RLock()
	defer rl.mu.RUnlock()

	paths := make([]string, 0, len(rl.skipPaths))
	for path := range rl.skipPaths {
		paths = append(paths, path)
	}
	return paths
}

// SetEndpointLimit dynamically sets a limit for a specific endpoint
func (rl *RateLimiter) SetEndpointLimit(path string, limit int) {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	if rl.endpointLimits == nil {
		rl.endpointLimits = make(map[string]int)
	}
	rl.endpointLimits[path] = limit
}

// GetStats returns current statistics for debugging/monitoring
func (rl *RateLimiter) GetStats() map[string]interface{} {
	rl.mu.RLock()
	defer rl.mu.RUnlock()

	return map[string]interface{}{
		"default_limit":   rl.defaultLimit,
		"window_size":     rl.windowSize.String(),
		"endpoint_limits": rl.endpointLimits,
		"skip_paths":      rl.SkipPaths(),
		"tracked_clients": len(rl.clients),
	}
}

File: forge-shared/forge_shared/middleware/rate_limit.py
"""
Rate limiting middleware using Redis with sliding window algorithm.

Provides token bucket rate limiting for API endpoints with configurable
limits and sliding window counters. Supports:
- Per-IP rate limits
- Per-API-key rate limits
- Endpoint-specific limits with glob patterns
- Sliding window algorithm
- Redis-backed storage with in-memory fallback

Example:
    ```python
    from fastapi import FastAPI
    from forge_shared.middleware import RateLimitMiddleware, RateLimitConfig

    # Basic usage
    app = FastAPI()
    app.add_middleware(
        RateLimitMiddleware,
        redis_url="redis://localhost:6379",
        requests_per_minute=60
    )

    # With endpoint-specific limits
    config = RateLimitConfig(
        default_limit=100,  # requests per hour
        endpoint_limits={
            "/api/auth/login": "5/minute",
            "/api/auth/signup": "10/hour",
            "/api/admin/*": "20/minute",
        }
    )
    app.add_middleware(RateLimitMiddleware, config=config, redis_url="redis://localhost:6379")
    ```
"""

import logging
import time
from dataclasses import dataclass, field
from fnmatch import fnmatch
from typing import Any

from fastapi import Request, Response, status
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
from starlette.types import ASGIApp

logger = logging.getLogger(__name__)

# Type alias for Redis client (lazy loaded)
RedisClient = Any


def _get_redis_module():
    """Lazy import redis module to avoid import errors when not using Redis."""
    try:
        import redis as redis_module

        return redis_module
    except ImportError:
        logger.warning("Redis package not installed. Rate limiting will use in-memory fallback.")
        return None


@dataclass
class RateLimitConfig:
    """Configuration for rate limiting.

    Attributes:
        default_limit: Default requests per window (default: 100/hour)
        default_window_seconds: Default window size in seconds (default: 3600)
        endpoint_limits: Dict mapping path patterns to rate limit strings
        exempt_paths: Paths exempt from rate limiting
        trusted_proxies: Trusted proxy networks (CIDR notation)
        enable_api_key_limits: Enable per-API-key rate limiting via X-API-Key header
        key_prefix: Redis key prefix for rate limit counters
    """

    default_limit: int = 100
    default_window_seconds: int = 3600  # 1 hour
    endpoint_limits: dict[str, str] = field(default_factory=dict)
    exempt_paths: list = field(
        default_factory=lambda: ["/health", "/docs", "/openapi.json", "/metrics"]
    )
    trusted_proxies: list = field(default_factory=list)
    enable_api_key_limits: bool = True
    key_prefix: str = "rate_limit"


def parse_rate_limit(limit_str: str) -> tuple[int, int]:
    """Parse rate limit string like '5/minute' or '100/hour'.

    Args:
        limit_str: Rate limit string in format "count/period"

    Returns:
        Tuple of (count, window_seconds)

    Examples:
        >>> parse_rate_limit("5/minute")
        (5, 60)
        >>> parse_rate_limit("100/hour")
        (100, 3600)
    """
    parts = limit_str.lower().split("/")
    if len(parts) != 2:
        return 100, 3600  # Default: 100/hour

    try:
        count = int(parts[0])
    except ValueError:
        return 100, 3600

    period = parts[1]

    period_map = {
        "second": 1,
        "seconds": 1,
        "s": 1,
        "minute": 60,
        "minutes": 60,
        "m": 60,
        "hour": 3600,
        "hours": 3600,
        "h": 3600,
        "day": 86400,
        "days": 86400,
        "d": 86400,
    }

    window = period_map.get(period, 3600)
    return count, window


class InMemoryRateLimiter:
    """In-memory rate limiter for fallback when Redis is unavailable.

    Uses a simple sliding window with cleanup on each check.
    Not suitable for multi-instance deployments.
    """

    def __init__(self):
        self._store: dict[str, list] = {}
        self._last_cleanup = time.time()

    def check_and_increment(
        self, key: str, limit: int, window_seconds: int
    ) -> tuple[bool, int, float]:
        """Check rate limit and increment counter.

        Args:
            key: Unique identifier for the client
            limit: Maximum requests allowed
            window_seconds: Window size in seconds

        Returns:
            Tuple of (allowed, remaining, retry_after)
        """
        current_time = time.time()
        window_start = current_time - window_seconds

        # Periodic cleanup (every 60 seconds)
        if current_time - self._last_cleanup > 60:
            self._cleanup(current_time)
            self._last_cleanup = current_time

        # Get or create entry
        if key not in self._store:
            self._store[key] = []

        # Remove expired entries
        self._store[key] = [t for t in self._store[key] if t > window_start]

        # Check limit
        current_count = len(self._store[key])

        if current_count >= limit:
            # Calculate retry after
            if self._store[key]:
                retry_after = self._store[key][0] + window_seconds - current_time
            else:
                retry_after = window_seconds
            return False, 0, max(0, retry_after)

        # Increment
        self._store[key].append(current_time)
        return True, limit - current_count - 1, 0

    def _cleanup(self, current_time: float, max_age: int = 86400):
        """Remove entries older than max_age seconds."""
        cutoff = current_time - max_age
        keys_to_remove = []
        for key, timestamps in self._store.items():
            self._store[key] = [t for t in timestamps if t > cutoff]
            if not self._store[key]:
                keys_to_remove.append(key)
        for key in keys_to_remove:
            del self._store[key]


class RateLimitMiddleware(BaseHTTPMiddleware):
    """
    Rate limiting middleware with Redis backend and in-memory fallback.

    Implements sliding window algorithm with:
    - Endpoint-specific limits via glob patterns
    - Per-API-key rate limiting via X-API-Key header
    - Per-IP rate limiting as fallback
    - Redis-backed storage with automatic in-memory fallback

    Attributes:
        app: ASGI application
        redis_url: Redis connection URL (None for in-memory only)
        config: RateLimitConfig instance
        requests_per_minute: Legacy parameter for backward compatibility

    Example:
        ```python
        # Basic usage
        app.add_middleware(
            RateLimitMiddleware,
            redis_url="redis://localhost:6379",
            requests_per_minute=60
        )

        # With endpoint-specific limits
        config = RateLimitConfig(
            endpoint_limits={
                "/api/auth/login": "5/minute",
                "/api/admin/*": "20/minute",
            }
        )
        app.add_middleware(RateLimitMiddleware, config=config, redis_url="redis://localhost:6379")
        ```
    """

    def __init__(
        self,
        app: ASGIApp,
        redis_url: str | None = None,
        requests_per_minute: int = 60,
        burst_size: int = 10,
        key_prefix: str = "rate_limit",
        config: RateLimitConfig | None = None,
    ) -> None:
        """
        Initialize rate limiting middleware.

        Args:
            app: ASGI application
            redis_url: Redis connection URL (None to use in-memory only)
            requests_per_minute: Maximum requests per minute (legacy, use config instead)
            burst_size: Burst capacity (legacy, not used in sliding window)
            key_prefix: Redis key prefix (legacy, use config instead)
            config: RateLimitConfig instance for full configuration
        """
        super().__init__(app)
        self.redis_url = redis_url
        self._redis: RedisClient | None = None
        self._redis_module = None

        # Use config if provided, otherwise create from legacy params
        if config is None:
            config = RateLimitConfig(
                default_limit=requests_per_minute,
                default_window_seconds=60,
                key_prefix=key_prefix,
            )
        self.config = config

        # In-memory fallback
        self._memory_limiter = InMemoryRateLimiter()

    async def _get_redis(self) -> RedisClient | None:
        """Get or create Redis connection."""
        if self._redis is not None:
            return self._redis

        if self.redis_url is None:
            return None

        # Lazy import redis
        if self._redis_module is None:
            self._redis_module = _get_redis_module()

        if self._redis_module is None:
            logger.warning("Redis not available, using in-memory rate limiting")
            return None

        try:
            self._redis = self._redis_module.from_url(self.redis_url, decode_responses=True)
            return self._redis
        except Exception as e:
            logger.error(f"Failed to connect to Redis: {e}")
            return None

    async def dispatch(self, request: Request, call_next) -> Response:
        """
        Process request with rate limiting.

        Args:
            request: Incoming request
            call_next: Next middleware or route handler

        Returns:
            Response or 429 Too Many Requests
        """
        path = request.url.path

        # Check exempt paths
        if path in self.config.exempt_paths:
            return await call_next(request)

        # Get rate limit for this endpoint
        limit, window = self._get_endpoint_limit(path)

        # Get client identifier
        identifier = self._get_identifier(request)
        key = f"{self.config.key_prefix}:{identifier}:{path}"

        # Check rate limit
        allowed, remaining, retry_after = await self._check_rate_limit(key, limit, window)

        if not allowed:
            return JSONResponse(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                content={
                    "error": "Too many requests",
                    "retry_after": int(retry_after),
                    "limit": limit,
                    "window_seconds": window,
                },
                headers={
                    "Retry-After": str(int(retry_after)),
                    "X-RateLimit-Limit": str(limit),
                    "X-RateLimit-Remaining": "0",
                    "X-RateLimit-Reset": str(int(time.time() + retry_after)),
                },
            )

        # Process request
        response = await call_next(request)

        # Add rate limit headers
        response.headers["X-RateLimit-Limit"] = str(limit)
        response.headers["X-RateLimit-Remaining"] = str(remaining)
        response.headers["X-RateLimit-Reset"] = str(int(time.time() + window))

        return response

    def _get_endpoint_limit(self, path: str) -> tuple[int, int]:
        """Get rate limit for a specific endpoint.

        Matches against glob patterns in endpoint_limits.

        Args:
            path: Request path

        Returns:
            Tuple of (limit, window_seconds)
        """
        # Check for exact match first
        if path in self.config.endpoint_limits:
            return parse_rate_limit(self.config.endpoint_limits[path])

        # Check for glob pattern matches
        for pattern, limit_str in self.config.endpoint_limits.items():
            if fnmatch(path, pattern):
                return parse_rate_limit(limit_str)

        # Return default
        return self.config.default_limit, self.config.default_window_seconds

    def _get_identifier(self, request: Request) -> str:
        """
        Get client identifier for rate limiting.

        Priority:
        1. API key from X-API-Key header (if enabled)
        2. User ID from request.state.user (if authenticated)
        3. IP address (fallback)

        Args:
            request: Request object

        Returns:
            Client identifier string
        """
        # Check for API key
        if self.config.enable_api_key_limits:
            api_key = request.headers.get("X-API-Key")
            if api_key:
                return f"apikey:{api_key[:16]}"  # Truncate for privacy

        # Use user ID if authenticated
        user = getattr(request.state, "user", None)
        if user:
            user_id = getattr(user, "id", None) or getattr(user, "user_id", None)
            if user_id:
                return f"user:{user_id}"

        # Fall back to IP address
        ip = self._get_client_ip(request)
        return f"ip:{ip}"

    def _get_client_ip(self, request: Request) -> str:
        """Get client IP address, respecting X-Forwarded-For.

        Args:
            request: Request object

        Returns:
            Client IP address
        """
        # Check X-Forwarded-For header
        forwarded = request.headers.get("X-Forwarded-For")
        if forwarded:
            # Take the first IP (original client)
            return forwarded.split(",")[0].strip()

        # Fall back to direct client IP
        if request.client:
            return request.client.host

        return "unknown"

    async def _check_rate_limit(
        self, key: str, limit: int, window_seconds: int
    ) -> tuple[bool, int, float]:
        """
        Check if request is within rate limit.

        Args:
            key: Redis key for this client/endpoint
            limit: Maximum requests allowed
            window_seconds: Window size in seconds

        Returns:
            Tuple of (allowed, remaining, retry_after)
        """
        redis_client = await self._get_redis()

        if redis_client is None:
            # Use in-memory fallback
            return self._memory_limiter.check_and_increment(key, limit, window_seconds)

        return await self._check_rate_limit_redis(redis_client, key, limit, window_seconds)

    async def _check_rate_limit_redis(
        self, redis_client: RedisClient, key: str, limit: int, window_seconds: int
    ) -> tuple[bool, int, float]:
        """Check rate limit using Redis sliding window.

        Args:
            redis_client: Redis client instance
            key: Redis key for this client/endpoint
            limit: Maximum requests allowed
            window_seconds: Window size in seconds

        Returns:
            Tuple of (allowed, remaining, retry_after)
        """
        current_time = int(time.time())
        window_start = current_time - window_seconds

        try:
            # Clean old entries (sliding window)
            await redis_client.zremrangebyscore(key, 0, window_start)

            # Count requests in current window
            count = await redis_client.zcard(key)

            if count >= limit:
                # Rate limit exceeded
                oldest = await redis_client.zrange(key, 0, 0, withscores=True)
                if oldest:
                    retry_after = oldest[0][1] + window_seconds - current_time
                else:
                    retry_after = window_seconds
                return False, 0, max(0, retry_after)

            # Add current request
            await redis_client.zadd(key, {str(current_time): current_time})
            await redis_client.expire(key, window_seconds)

            return True, limit - count - 1, 0

        except Exception as e:
            logger.error(f"Redis rate limit error: {e}")
            # Fall back to in-memory
            return self._memory_limiter.check_and_increment(key, limit, window_seconds)


async def close_redis() -> None:
    """
    Close Redis connection pool.

    Should be called on application shutdown.
    """
    # This would be implemented with proper connection pool management
    pass


# Convenience function for creating rate limit configs
def create_rate_limits(
    default: str = "100/hour",
    auth: str = "5/minute",
    admin: str = "20/minute",
    public: str = "1000/hour",
) -> RateLimitConfig:
    """Create a standard rate limit configuration.

    Args:
        default: Default rate limit
        auth: Rate limit for auth endpoints
        admin: Rate limit for admin endpoints
        public: Rate limit for public endpoints

    Returns:
        RateLimitConfig instance

    Example:
        ```python
        config = create_rate_limits(
            default="100/hour",
            auth="5/minute",
            admin="20/minute",
            public="1000/hour"
        )
        app.add_middleware(RateLimitMiddleware, config=config, redis_url="redis://localhost:6379")
        ```
    """
    default_limit, default_window = parse_rate_limit(default)

    return RateLimitConfig(
        default_limit=default_limit,
        default_window_seconds=default_window,
        endpoint_limits={
            "/api/auth/login": auth,
            "/api/auth/signup": auth,
            "/api/auth/*": auth,
            "/api/admin/*": admin,
            "/api/public/*": public,
        },
    )
